File: app/src/main/java/com/ucsdextandroid1/kotlinlist/SongViewHolder.java
package com.ucsdextandroid1.kotlinlist;

import android.graphics.drawable.ColorDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;
import com.ucsdextandroid1.kotlinlist.Models.SongItem;

/**
 * Created by rjaylward on 2019-06-08
 */
public class SongViewHolder extends RecyclerView.ViewHolder {

    private ImageView imageView = itemView.findViewById(R.id.vsi_image);
    private TextView titleView = itemView.findViewById(R.id.vsi_title);
    private TextView subtitleView;

    private SongItem currentSongItem;
    private OnItemClickListener<SongItem> clickListener;

    public static SongViewHolder inflate(@NonNull ViewGroup parent) {
        //TODO replace with extension method parent.inflate(...)
        return new SongViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.viewholder_song_item, parent, false));
    }

    private SongViewHolder(@NonNull View itemView) {
        super(itemView);

        subtitleView = itemView.findViewById(R.id.vsi_subtitle);

        itemView.setOnClickListener(v -> {
            if(clickListener != null && currentSongItem != null)
                clickListener.onItemClicked(currentSongItem);
        });
    }

    void bind(SongItem songItem) {
        this.currentSongItem = songItem;

        Picasso.get().load(songItem.getArtworkUrl())
                .placeholder(new ColorDrawable(Utils.randomColor()))
                .into(imageView);

        titleView.setText(songItem.getTrackName());
        subtitleView.setText(String.format("%s • %s", songItem.getAlbumName(), songItem.getArtistName()));
    }

    void setClickListener(OnItemClickListener<SongItem> clickListener) {
        this.clickListener = clickListener;
    }

}

File: app/src/main/java/com/ucsdextandroid1/kotlinlist/SongViewHolder.kt
package com.ucsdextandroid1.kotlinlist

import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.squareup.picasso.Picasso
import com.ucsdextandroid1.kotlinlist.Models.SongItem

/**
 * Created by rjaylward on 2019-06-08
 */
class SongViewHolder private constructor(itemView: View): RecyclerView.ViewHolder(itemView) {

    private val imageView: ImageView = itemView.findViewById(R.id.vsi_image)
    private val titleView: TextView = itemView.findViewById(R.id.vsi_title)
    private val subtitleView: TextView = itemView.findViewById(R.id.vsi_subtitle)

    private var currentSongItem: SongItem? = null

    var clickListener: OnItemClickListener<SongItem>? = null

    init {
        itemView.setOnClickListener {
            if(currentSongItem != null)
                clickListener?.onItemClicked(currentSongItem)
        }
    }

    fun bind(songItem: SongItem) {
        this.currentSongItem = songItem

        imageView.loadImageUrl(songItem.artworkUrl)

        titleView.text = songItem.trackName
        subtitleView.text = String.format("%s • %s", songItem.albumName, songItem.artistName)
    }

    companion object {

        @JvmStatic
        fun inflate(parent: ViewGroup): SongViewHolder {
            return SongViewHolder(parent.inflate(R.layout.viewholder_song_item))
        }

    }

}
